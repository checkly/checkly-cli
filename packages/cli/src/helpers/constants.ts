export const DEFAULT_REGION = 'eu-central-1'
