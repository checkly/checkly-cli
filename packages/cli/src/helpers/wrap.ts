export interface WrapOptions {
  length?: number
  prefix?: string
  trimPrefix?: boolean
}

export function wrap (
  content: string,
  { length = 80, prefix = '', trimPrefix = true }: WrapOptions = {},
): string {
  const tokens = content.split(/([ \t\r\n])/)
  const lines = []
  let line = ''
  let trim = 0

  const prefixTrim = trimPrefix
    ? (prefix.match(/[ \t]$/) ?? [''])[0].length
    : 0

  for (const token of tokens) {
    if (token === '') {
      // Between tokens, not useful, skip.
      continue
    }

    if (token === '\r') {
      // Remove entirely.
      continue
    }

    if (token === ' ' || token === '\t') {
      // Add indiscriminately. These will get trimmed if needed.
      line += token
      trim += 1
      continue
    }

    if (token === '\n') {
      // Respect line endings.
      lines.push(line.slice(0, line.length - trim))
      line = prefix
      trim = prefixTrim
      continue
    }

    if (line.length === 0) {
      // We can assume that the first token of a line will always fit. Even
      // if it does not, it would not fit on any other line either.
      line += prefix
      line += token
      continue
    }

    if (line.length + token.length > length) {
      // Line would get too long, start a new line.
      lines.push(line.slice(0, line.length - trim))
      line = prefix
    }

    trim = 0
    line += token
  }

  lines.push(line.slice(0, line.length - trim))

  return lines.join('\n')
}
