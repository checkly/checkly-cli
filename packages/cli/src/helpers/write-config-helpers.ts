import * as recast from 'recast'
import path from 'node:path'
import fs from 'node:fs/promises'
import { namedTypes as n } from 'ast-types';


export function findPropertyByName (ast: any, name: string): recast.types.namedTypes.Property | undefined {
  let node
  recast.visit(ast, {
    visitProperty (path: any) {
      if (path.node.key.name === name) {
        node = path.node
      }
      return false
    },
  })
  return node
}

export function addOrReplaceItem(ast: any, node: any, name: string) {
  const item = findPropertyByName(ast, name)
  if (item) {
    item.value = node.value
  } else {
    ast.properties.push(node)
  }
}

export function addItemToArray(ast: any, node: any, name: string) {
  if (!n.ObjectExpression.check(ast)) {
    throw new Error('AST node is not an ObjectExpression');
  }

  const item = findPropertyByName(ast, name);

  if (item) {
    if (n.ArrayExpression.check(item.value)) {
      item.value.elements = item.value.elements || [];
      item.value.elements.push(node);
    } else {
      throw new Error(`Property "${name}" exists but is not an array`);
    }
  } else {
    ast.properties.push({
      type: 'Property',
      key: { type: 'Identifier', name },
      value: { type: 'ArrayExpression', elements: [node] },
      kind: 'init',
      computed: false,
      method: false,
      shorthand: false,
    });
  }
}


export async function reWriteChecklyConfigFile (data: string, fileName: string, dir: string) {
  await fs.writeFile(path.join(dir, fileName), data)
}
