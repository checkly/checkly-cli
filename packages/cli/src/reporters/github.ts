import * as fs from 'fs'
import * as path from 'path'

import AbstractListReporter, { checkFilesMap } from './abstract-list'
import { formatDuration, printLn, getTestSessionUrl, getTraceUrl } from './util'

const outputFile = './checkly-github-report.md'

type GithubMdBuilderOptions = {
  testSessionId?: string
  testResultIds?: { [key: string]: string }
  numChecks: number
  runLocation: string
  checkFilesMap: checkFilesMap
}

export class GithubMdBuilder {
  testSessionId?: string
  testResultIds?: { [key: string]: string }
  numChecks: number
  runLocation: string
  checkFilesMap: checkFilesMap
  subHeader: Array<string>
  tableHeaders: Array<string>
  extraTableHeadersWithLinks: Array<string>
  tableRows: Array<string> = []

  readonly header: string = '# Checkly Test Session Summary'
  readonly tableSeparatorFiller: string = '|:-'
  readonly tableSeparator: string = '|'
  constructor (options: GithubMdBuilderOptions) {
    this.testSessionId = options.testSessionId
    this.testResultIds = options.testResultIds
    this.numChecks = options.numChecks
    this.runLocation = options.runLocation
    this.checkFilesMap = options.checkFilesMap

    this.subHeader = []
    this.tableHeaders = ['Result', 'Name', 'Check Type', 'Filename', 'Duration']
    this.extraTableHeadersWithLinks = ['Assets', 'Link']
    this.tableRows = []
  }

  render (): string {
    this.subHeader.push(`Ran **${this.numChecks}** checks in **${this.runLocation}**.`)

    if (this.testSessionId) {
      this.subHeader.push(`[View detailed test session summary](${getTestSessionUrl(this.testSessionId)})`)
    }

    if (this.testSessionId && this.testResultIds) {
      this.tableHeaders = this.tableHeaders.concat(this.extraTableHeadersWithLinks)
    }

    for (const [_, checkMap] of this.checkFilesMap.entries()) {
      for (const [_, { check, result }] of checkMap.entries()) {
        const tableRow: Array<string> = [
          `${result.hasFailures ? '❌ Fail' : '✅ Pass'}`,
          `${result.name}`,
          `${result.checkType}`,
          `\`${result.sourceFile}\``,
          `${formatDuration(result.responseTime)} `,
        ]

        if (this.testSessionId && this.testResultIds) {
          const assets: Array<string> = []

          if (result.hasFailures && result.traceFilesUrls) {
            assets.push(`[Trace](${getTraceUrl(result.traceFilesUrls[0])})`)
          }

          if (result.hasFailures && result.videoFilesUrls) {
            assets.push(`[Video](${result.videoFilesUrls[0]})`)
          }

          const assetsColumn: string = assets.join(' \\| ')

          const linkColumn = `[Full test report](${getTestSessionUrl(this.testSessionId)}/results/${this.testResultIds[result.logicalId]})`
          tableRow.push(assetsColumn, linkColumn)
        }

        this.tableRows.push(this.tableSeparator + tableRow.join(this.tableSeparator) + this.tableSeparator)
      }
    }

    let markdown = this.header + '\n' +
    this.subHeader.join('\n') + '\n' +
    this.tableSeparator + this.tableHeaders.join('|') + this.tableSeparator + '\n' +
    this.tableSeparatorFiller.repeat(this.tableHeaders.length) + this.tableSeparator + '\n' +
    this.tableRows.sort((a, b) => a < b ? 1 : -1).join('\n') + '\n'

    if (!this.testSessionId) {
      markdown = markdown + '> Tip: use `--record` to get a full test session overview with traces, videos and logs, e.g. `npx checkly test --reporter=github --record`'
    }

    return markdown
  }
}

export default class GithubReporter extends AbstractListReporter {
  onBeginStatic () {
    printLn(`Running ${this.numChecks} checks in ${this._runLocationString()}.`, 2, 1)
  }

  onBegin (testSessionId?: string, testResultIds?: { [key: string]: string }) {
    this.onBeginStatic()
    this._setTestSessionId(testSessionId)
    this._setTestResultIds(testResultIds)
  }

  onEnd () {
    this._printBriefSummary()
    const githubMdBuilder = new GithubMdBuilder({
      testSessionId: this.testSessionId,
      testResultIds: this.testResultIds,
      numChecks: this.numChecks,
      runLocation: this._runLocationString(),
      checkFilesMap: this.checkFilesMap,
    })

    const markDown = githubMdBuilder.render()

    const summaryFilename = process.env.CHECKLY_REPORTER_GITHUB_OUTPUT ?? outputFile
    fs.mkdirSync(path.resolve(path.dirname(summaryFilename)), { recursive: true })
    fs.writeFileSync(summaryFilename, markDown)

    printLn(`Github summary saved in '${path.resolve(summaryFilename)}'.`, 2)
  }
}
