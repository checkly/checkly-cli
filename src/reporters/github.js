class GithubReporter {
  // TODO
}

module.exports = GithubReporter
