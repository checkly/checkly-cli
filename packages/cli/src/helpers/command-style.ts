import chalk from 'chalk'
import { ux } from '@oclif/core'

import { BaseCommand } from '../commands/baseCommand'
import { wrap } from './wrap'
import logSymbols from 'log-symbols'

const wrapOptions = {
  prefix: '  ',
  length: 78,
}

export class CommandStyle {
  private c: BaseCommand

  constructor (command: BaseCommand) {
    this.c = command
  }

  actionStart (message: string) {
    if (this.c.fancy) {
      ux.action.start(message, undefined, { stdout: true })
    }
  }

  actionSuccess () {
    if (this.c.fancy) {
      ux.action.stop(`✅`)
      this.c.log()
    }
  }

  actionFailure () {
    if (this.c.fancy) {
      ux.action.stop(`❌`)
      this.c.log()
    }
  }

  comment (message: string) {
    this.c.log(chalk.cyan(wrap(message, { ...wrapOptions, prefix: '// ' })))
    this.c.log()
  }

  longSuccess (title: string, message: string) {
    this.c.log(`${logSymbols.success} ${title}`)
    this.c.log()
    this.c.log(wrap(message, wrapOptions))
    this.c.log()
  }

  shortSuccess (message: string) {
    this.c.log(`${logSymbols.success} ${message}`)
    this.c.log()
  }

  longInfo (title: string, message: string) {
    this.c.log(`${logSymbols.info} ${title}`)
    this.c.log()
    this.c.log(wrap(message, wrapOptions))
    this.c.log()
  }

  shortInfo (message: string) {
    this.c.log(`${logSymbols.info} ${message}`)
    this.c.log()
  }

  longWarning (title: string, message: string) {
    this.c.log(`${logSymbols.warning} ${title}`)
    this.c.log()
    this.c.log(chalk.yellow(wrap(message, wrapOptions)))
    this.c.log()
  }

  shortWarning (message: string) {
    this.c.log(`${logSymbols.warning} ${chalk.yellow(message)}`)
    this.c.log()
  }

  longError (title: string, message: string) {
    this.c.log(`${logSymbols.error} ${title}`)
    this.c.log()
    this.c.log(chalk.red(wrap(message, wrapOptions)))
    this.c.log()
  }

  shortError (message: string) {
    this.c.log(`${logSymbols.error} ${chalk.red(message)}`)
    this.c.log()
  }

  fatal (message: string) {
    this.c.log(chalk.red(message))
    this.c.log()
  }
}
