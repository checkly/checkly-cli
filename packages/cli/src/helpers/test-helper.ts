import { PrivateRunLocation, RunLocation } from '../services/abstract-check-runner'
import { Session } from '../constructs'
import type { Region } from '..'
import { ReporterType } from '../reporters/reporter'
import { isCI } from 'ci-info'

const DEFAULT_REGION = 'eu-central-1'

export async function prepareRunLocation (
  configOptions: { runLocation?: keyof Region, privateRunLocation?: string } = {},
  cliFlags: { runLocation?: keyof Region, privateRunLocation?: string } = {},
  api: any,
  accountId: string,
): Promise<RunLocation> {
  // Command line options take precedence
  if (cliFlags.runLocation) {
    const { data: availableLocations } = await api.locations.getAll()
    if (availableLocations.some((l: { region: string | undefined }) => l.region === cliFlags.runLocation)) {
      return { type: 'PUBLIC', region: cliFlags.runLocation }
    }
    throw new Error(`Unable to run checks on unsupported location "${cliFlags.runLocation}". ` +
      `Supported locations are:\n${availableLocations.map((l: { region: any }) => `${l.region}`).join('\n')}`)
  } else if (cliFlags.privateRunLocation) {
    return preparePrivateRunLocation(cliFlags.privateRunLocation, api, accountId)
  } else if (configOptions.runLocation && configOptions.privateRunLocation) {
    throw new Error('Both runLocation and privateRunLocation fields were set in your Checkly config file.' +
      ` Please only specify one run location. The configured locations were' +
        ' "${configOptions.runLocation}" and "${configOptions.privateRunLocation}"`)
  } else if (configOptions.runLocation) {
    return { type: 'PUBLIC', region: configOptions.runLocation }
  } else if (configOptions.privateRunLocation) {
    return preparePrivateRunLocation(configOptions.privateRunLocation, api, accountId)
  } else {
    return { type: 'PUBLIC', region: DEFAULT_REGION }
  }
}

export async function preparePrivateRunLocation (privateLocationSlugName: string, api: any, accountId: string): Promise<PrivateRunLocation> {
  try {
    const privateLocations = await Session.getPrivateLocations()
    const privateLocation = privateLocations.find(({ slugName }) => slugName === privateLocationSlugName)
    if (privateLocation) {
      return { type: 'PRIVATE', id: privateLocation.id, slugName: privateLocationSlugName }
    }
    const { data: account } = await api.accounts.get(accountId)
    throw new Error(`The specified private location "${privateLocationSlugName}" was not found on account "${account.name}".`)
  } catch (err: any) {
    throw new Error(`Failed to get private locations. ${err.message}.`)
  }
}

export function prepareReportersTypes (reporterFlag: ReporterType, cliReporters: ReporterType[] = []): ReporterType[] {
  if (!reporterFlag && !cliReporters.length) {
    return [isCI ? 'ci' : 'list']
  }
  return reporterFlag ? [reporterFlag] : cliReporters
}
