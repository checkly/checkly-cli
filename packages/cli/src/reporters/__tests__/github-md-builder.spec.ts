import { GithubMdBuilder } from '../github'
import { generateMapAndTestResultIds } from './helpers'

const testSessionId = '0c4c64b3-79c5-44a6-ae07-b580ce73f328'
const runLocation = 'eu-west-1'
describe('GithubMdBuilder', () => {
  test('renders basic markdown output with no assets & links', () => {
    const { checkFilesMap } = generateMapAndTestResultIds()
    const markdown = new GithubMdBuilder({
      testSessionId: undefined,
      testResultIds: {},
      numChecks: checkFilesMap.size,
      runLocation,
      checkFilesMap,
    }).render()
    expect(markdown).toMatchSnapshot('github-basic-markdown')
  })
  test('renders basic markdown output with assets & links', () => {
    const { checkFilesMap, testResultIds } = generateMapAndTestResultIds()
    const markdown = new GithubMdBuilder({
      testSessionId,
      testResultIds,
      numChecks: checkFilesMap.size,
      runLocation,
      checkFilesMap,
    }).render()
    expect(markdown).toMatchSnapshot('github-markdown-with-assets-links')
  })
})
