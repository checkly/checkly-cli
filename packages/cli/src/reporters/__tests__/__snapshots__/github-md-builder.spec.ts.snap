// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GithubMdBuilder renders basic markdown output with assets & links: github-markdown-with-assets-links 1`] = `
"# Checkly Test Session Summary
Ran **2** checks in **eu-west-1**.
[View detailed test session summary](https://app-test.checklyhq.com/test-sessions/0c4c64b3-79c5-44a6-ae07-b580ce73f328)
|Result|Name|Check Type|Filename|Duration|Assets|Link|
|:-|:-|:-|:-|:-|:-|:-|
|✅ Pass|my-test.spec.ts|BROWSER|\`src/__checks__/folder/browser.check.ts\`|7s ||[Full test report](https://app-test.checklyhq.com/test-sessions/0c4c64b3-79c5-44a6-ae07-b580ce73f328/results/702961fd-7e2c-45f0-97be-1aa9eabd4d82)|
|✅ Pass|Test API check|API|\`src/some-other-folder/api.check.ts\`|2s ||[Full test report](https://app-test.checklyhq.com/test-sessions/0c4c64b3-79c5-44a6-ae07-b580ce73f328/results/1c0be612-a5ec-432e-ac1c-837d2f70c010)|
"
`;

exports[`GithubMdBuilder renders basic markdown output with no assets & links: github-basic-markdown 1`] = `
"# Checkly Test Session Summary
Ran **2** checks in **eu-west-1**.
|Result|Name|Check Type|Filename|Duration|
|:-|:-|:-|:-|:-|
|✅ Pass|my-test.spec.ts|BROWSER|\`src/__checks__/folder/browser.check.ts\`|7s |
|✅ Pass|Test API check|API|\`src/some-other-folder/api.check.ts\`|2s |
> Tip: use \`--record\` to get a full test session overview with traces, videos and logs, e.g. \`npx checkly test --reporter=github --record\`"
`;
